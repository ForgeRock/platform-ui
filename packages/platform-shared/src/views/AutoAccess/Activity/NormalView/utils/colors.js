/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

export const colorsHex = ['#2ed47a', '#ffb946', '#109cf1', '#f7685b', '#61e2e2'];
export const colorsReadable = ['green', 'yellow', 'blue', 'red', 'cyan'];
