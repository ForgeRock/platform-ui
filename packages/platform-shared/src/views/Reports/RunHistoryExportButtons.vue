<!-- Copyright (c) 2023 ForgeRock. All rights reserved.

This software may be modified and distributed under the terms
of the MIT license. See the LICENSE file for details. -->
<template>
  <div>
    <FrIcon
      v-if="props.exportStatus === 'export' || props.exportStatus === 'error'"
      data-testid="fr-export-button-icon"
      name="sync" />
    <BSpinner
      v-else-if="props.exportStatus === 'exporting' || props.exportStatus === 'downloading'"
      class="mr-1"
      data-testid="fr-export-button-spinner"
      small
      :label="$t('common.loadingEtc')" />
    <FrIcon
      v-else-if="props.exportStatus === 'download' || props.exportStatus === 'error'"
      data-testid="fr-download-button-icon"
      name="file_download" />
    <span data-testid="action-button-label">
      {{ props.label }}
    </span>
    <div class="fr-report-error-badge d-inline-block ml-2">
      <FrRunReportBadges
        v-if="props.exportStatus === 'error'"
        report-status="error" />
    </div>
  </div>
</template>

<script setup>
/**
 * @description Run History export buttons
 */
import { BSpinner } from 'bootstrap-vue';
import FrIcon from '@forgerock/platform-shared/src/components/Icon';
import FrRunReportBadges from './RunReportBadges';

const props = defineProps({
  exportStatus: {
    type: String,
    default: 'export',
    validator(value) {
      return ['export', 'exporting', 'download', 'downloading', 'error'].includes(value);
    },
  },
  fileType: {
    type: String,
    default: '',
  },
  label: {
    type: String,
    default: '',
  },
});
</script>

<style lang="scss" scoped>
  .fr-report-error-badge {
    width: 48px;
  }
</style>
