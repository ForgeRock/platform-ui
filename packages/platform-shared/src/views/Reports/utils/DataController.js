/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import {
  getAMTreesConfig,
  getCertificationsForReports,
  getGovernanceResource,
  getGovPolicyList,
  getGovPolicyRuleList,
  getOauth2Clients,
  getRequestTypesForReports,
  getTeamMembers,
  managedResourcePropertyRequest,
  requestTrees,
} from './ReportsApiHelper';

/**
 * @description Data controller object that maps the data source to the API
 * request. The deeper the level, the more priority it has over the parent
 * level. For example, if a parent level has a fetch function, but the child
 * level also has a fetch function, the child level fetch function will be used.
 *
 * { SERVICE } First Level is the datasource service name (AM, IDM, IGA).
 * { entity } Second Level is the data source entity name.
 * { attribute } Third Level is the data source entity property or attribute name.
 *
 * AVAILABLE PROPERTIES:
 * - fetch: The API request function to fetch the data.
 * - mutation: The function to transform the data.
 * - canQuery: Boolean to determine if the data can be queried.
 * - endUserCanFetch: Boolean to determine if the end user can fetch the data.
 * - taggable: Boolean to determine if the parameter field can be tagged.
 * - {*}: Additional properties can be added and will be injected into
 *   the parameter request object, which in-turn gets injected into the
 *   fetch function as the first argument.
 */

const DATA_CONTROLLER = Object.freeze({
  AM: {
    amconfig: {
      fetch: getAMTreesConfig,
      mutation: ({ suspendedAuthenticationTimeout: first, authenticationSessionsMaxDuration: second }) => Math.min(first, second),
    },
    trees: {
      fetch: requestTrees,
      canQuery: false,
    },
  },
  IDM: {
    fetch: managedResourcePropertyRequest,
    application: {},
    group: {},
    oauth2application: { fetch: getOauth2Clients },
    organization: {},
    role: {},
    teammember: { fetch: getTeamMembers },
    user: {},
  },
  IGA: {
    certification: { fetch: getCertificationsForReports },
    // Currently, IGA does not have the capability to query
    // the entitlements object directly, so the only way to
    // query entitlements is by "displayName" under the
    // "descriptor" object, and this is why "targetName"
    // and "sortBy" are hard-coded to point to "displayName".
    entitlement: {
      fetch: getGovernanceResource,
      mutation: (data, attribute) => data.map((obj) => obj.entitlement[attribute] || obj),
      body: (term = '') => ({
        targetFilter: {
          operator: 'CONTAINS',
          operand: {
            targetName: 'descriptor.idx./entitlement.displayName',
            targetValue: term,
          },
        },
      }),
      queryParameters: {
        pageNumber: 0,
        pageSize: 10,
        sortBy: 'descriptor.idx./entitlement.displayName',
        sortDir: 'asc',
      },
    },
    policy: { fetch: getGovPolicyList },
    request: { fetch: getRequestTypesForReports },
    rule: { fetch: getGovPolicyRuleList },
    violation: {},
  },
});

export default DATA_CONTROLLER;
