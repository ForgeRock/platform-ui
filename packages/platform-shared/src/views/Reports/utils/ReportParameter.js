/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

/**
 * @description Class that generates an analytics report parameter object.
 */

export default class ReportParameter {
  /**
   * @description Constructor for the ReportParameter.
   * @param {Object} args Object containing the parameter properties
   */
  constructor(args) {
    this._id = args._id;
    this._model = args.model;
    this.component = args.component;
    this.description = args.description;
    this.internalSearch = args.internalSearch;
    this.label = args.label;
    this.placeholder = args.label;
    this.taggable = args.taggable;
    this.testId = `fr-field-${args._id}`;
    this.type = args.type;

    if (args.request) {
      this.request = args.request;
      this.request.fetchRequestData = (term) => this.request.fetch(this.request, term);
      this.request.mutateRequestData = (data) => this.request.mutation(data, this.request.attribute);
    } else if (args.enums) {
      this.enums = ReportParameter.enumMutation(args.enums);
    } else if (args.type === 'multiselect' || args.type === 'select') {
      this.options = [];
    }
  }

  get model() {
    return this._model;
  }

  set model(value) {
    this._model = value;
  }

  /**
   * @description Method to handle the data for a select / multiselect field.
   * @param {Array} data Data handled by the dropdown options
   * @returns {Array} Filtered array if field type is select or multiselect
   */
  selectOptionsDataHandler(data) {
    if (data && this.model) {
      if (this.type === 'select') {
        // The select component requires a selected value to be part of the
        // overall list of options and if it is not present, the selected value
        // will not be displayed.  This can happen when we fetch data when
        // searching within a field and selecting a value that is not in the
        // original list. After selection, the field will query for the original
        // list, which will not contain the selected value, and will make the
        // selected value disappear. To prevent this, we add the selected value
        // to the beginning of the list if it is not included in the response.
        if (!data.includes(this.model)) return [this.model, ...data];
      }
      if (this.type === 'multiselect') {
        // Filters out any selected values from the search results
        return data.filter((item) => !this.model.includes(item));
      }
    }

    return data;
  }

  /**
   * @description Method to format the payload for the API request.
   * @returns {Object} containing the parameter payload
   */
  payload() {
    let payload = this.model;
    if (this.type === 'date') payload = ReportParameter.dateToISO(payload);
    return { [this._id]: payload };
  }

  /**
   * @description Setter for the select component options which
   * can be triggered on the "@tag" event on the <FrField> component.
   * It checks if the value is already present in the enums,
   * request, or options arrays and adds it if it is not present.
   * @param {String} val The value to be added to the dropdown options
   */
  selectOptionsSetter(val) {
    const { enums, request, options } = this;
    const enumsText = enums ? enums.map(({ text }) => text.toLowerCase()) : [];

    if (enums && !enumsText.includes(val.toLowerCase())) {
      enums.unshift(ReportParameter.enumFormat(val));
    } else if (request && !request.data.includes(val)) {
      request.data.unshift(val);
    } else if (options && !options.includes(val)) {
      this.options.unshift(val);
    }

    this.model = val;
  }

  /**
   * @description Method to convert a date string to ISO format.
   * @param {String} date date string from the datepicker component
   * @returns {String} ISO formatted date string
   */
  static dateToISO(date) {
    return Date.parse(date) ? new Date(date).toISOString() : date;
  }

  /**
   * @description Method to format the enums that are received from the API.
   * @param {String} value use for the label and value of the enum
   * @returns {Object} containing the text and value of the enum
   */
  static enumFormat(value) {
    return { text: value, value };
  }

  /**
   * @description Formats the enums array that is passed to the FrField component.
   * @param {Array} enums Array of enums to be formatted
   * @returns {Array} containing the text and value of the enums
   */
  static enumMutation(enums) {
    return enums.map(({ name: text, value }) => ({ text, value }));
  }
}
