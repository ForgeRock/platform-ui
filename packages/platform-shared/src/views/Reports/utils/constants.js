/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

/**
 * Report viewer group name
 */
export const REPORT_VIEWER_GROUP = 'report_viewer';

/**
 * Used for mapping field type names that come from the API,
 * to the field type names that the <FrField> component consumes.
 */
export const FIELD_TYPE_MAP = {
  array: 'multiselect',
  boolean: 'boolean',
  date: 'date',
  float: 'number',
  integer: 'number',
  string: 'string',
};
