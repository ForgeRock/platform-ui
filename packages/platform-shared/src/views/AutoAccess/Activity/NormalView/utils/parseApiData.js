/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { colorsHex, colorsReadable } from './colors';

export default function parseApiData(response) {
  const categories = {};
  const categoryHasOther = {};
  const categoryNames = [...new Set(response.result.map((item) => item.category))];
  categoryNames.forEach((category) => {
    let categoryValue = {};
    let otherCount = 0;
    let otherCategoryData = {};
    categories[category] = [];
    categoryHasOther[category] = false;

    /**
     * This reorganizing of the API data is needed since the data provided is formatted in a way needed by the UI.
     *
     * The data is flat and need to be organized into a usable object structure.
     * It needs to be manipulated by merging values of 'other', and null together into a single field.
     * The field of 'Others' also need to be converted to have the label 'other' and be sorted last
     */
    response.result.forEach((item) => {
      if (item.category === category) {
        // 'other' or null
        if (item.value === 'other' || !item.value) {
          otherCount += parseInt(item.count, 10);
          if (!categoryHasOther[category]) {
            otherCategoryData = {
              label: 'unknown',
              value: otherCount,
              valueFormatted: otherCount,
            };
            categoryHasOther[category] = true;
          } else {
            otherCategoryData.value = otherCount;
            otherCategoryData.valueFormatted = otherCount;
          }
        } else {
          // Handle all normal categories, except 'Others' which gets renames to 'other'
          categoryValue = {
            label: item.value === 'Others' ? 'other' : item.value,
            value: parseInt(item.count, 10),
            valueFormatted: item.count,
          };
          categories[item.category].push(categoryValue);
        }
      }
    });

    // We have found 'other' or null and need to add the manipulated field to the main array
    if (categoryHasOther[category]) {
      categories[category].push(otherCategoryData);
    }

    // Sort categories based on numerical 'value', but leave 'other' sorted last
    Object.keys(categories).forEach((cat) => {
      categories[cat].sort((a, b) => {
        const aVal = parseInt(a.value, 10);
        const bVal = parseInt(b.value, 10);
        if (a.label === 'other') { return 1; }
        if (b.label === 'other') { return -1; }
        if (aVal < bVal) { return 1; }
        if (aVal > bVal) { return -1; }
        return 0;
      });

      // Set the color values - hex for the Donut Chart, readable for the data table
      let colorIndex = 0;
      categories[cat].forEach((item) => {
        item.color = colorsHex[colorIndex] || colorsHex[4];
        item.colorReadable = colorsReadable[colorIndex] || colorsReadable[4];
        colorIndex += 1;
      });
    });
  });

  return categories;
}
