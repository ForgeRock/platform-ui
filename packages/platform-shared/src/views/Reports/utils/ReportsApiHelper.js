/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { showErrorMessage } from '@forgerock/platform-shared/src/utils/notification';
import { getConfig, getAMConfig } from '@forgerock/platform-shared/src/api/ConfigApi';
import { getManagedResourceList } from '@forgerock/platform-shared/src/api/ManagedResourceApi';
import { getReportRuns, reportExportRequest } from '@forgerock/platform-shared/src/api/AutoApi';
import { actionGetAllTrees } from '@forgerock/platform-shared/src/api/TreeApi';
import { getGatewaysOrAgents } from '@forgerock/platform-shared/src/api/AgentsApi';
import { searchGovernanceResource } from '@forgerock/platform-shared/src/api/governance/CommonsApi';
import { getPolicyList, getPolicyRuleList } from '@forgerock/platform-shared/src/api/governance/PolicyApi';
import { getCertificationTemplates } from '@forgerock/platform-shared/src/api/governance/CertificationApi';
import { getRequestTypes } from '@forgerock/platform-shared/src/api/governance/AccessRequestApi';

import i18n from '@forgerock/platform-shared/src/i18n';
import store from '@/store';

/**
 * Gets a list of only oAuth2 clients
 * @param {Undefined} _ Unused first parameter
 * @param {String} term The search term to query.
 * @returns {Array} list of all oAuth2 clients
 */
export async function getOauth2Clients(_, term = '') {
  try {
    const response = await getGatewaysOrAgents('oauth2', { queryFilter: term }, 10);
    return response.data.result || [];
  } catch {
    return [];
  }
}

/**
 * Gets tenant admins
 * @param {Object} obj An object for the request query strings.
 * @param {String} obj.entity The entity to query.
 * @param {String} obj.attribute The attribute to query.
 * @param {String} term The search term to query.
 * @returns {Array} list of all team members
 */
export async function getTeamMembers({ entity, attribute }, term) {
  const queryFilter = term ? `${attribute} sw "${term}"` : true;
  try {
    const { data } = await getManagedResourceList(entity, { queryFilter, pageSize: 10, fields: attribute });
    return data.result || [];
  } catch {
    return [];
  }
}

/**
 * Requests a report export file or a download
 * @param {String} runId Report job run ID
 * @param {String} action Action to execute, can be 'export' or 'download'.
 * @param {String} template Report name
 * @param {String} format File type: jsonl || csv
 */
export async function requestExport(runId, action, template, format) {
  try {
    const response = await reportExportRequest(runId, {
      _action: action,
      name: template,
      format,
    });
    return response;
  } catch (error) {
    if (action === 'download') {
      showErrorMessage(error, i18n.global.t('reports.tabs.runHistory.errors.errorDownload'));
    }
    return error.name;
  }
}

/**
 * Gets report runs for the given report template name.
 * For all runs: { name }
 * For a single run: { runId, name }
 * @param {Object} params query params
 * @returns {Promise<Array>}
 */
export async function requestReportRuns(params) {
  try {
    const { result: reportRunsData } = await getReportRuns(params);
    return reportRunsData || [];
  } catch (error) {
    showErrorMessage(error, i18n.global.t('reports.tabs.runHistory.errors.errorRetrievingHistory'));
    return [];
  }
}

/**
 * Gets the supplied managed object from the config
 *
 * @param {String} managedObjectName managed object name that comes from the _FIELD_MAP config
 * @returns {Object} managed object
 */
async function getManagedObject(managedObjectName) {
  function findManagedObject(objects, name) {
    return objects.find((object) => (!store.state.isFraas || object.name.startsWith(store.state.realm)) && object.name.endsWith(name));
  }

  try {
    const { data: managedConfig } = await getConfig('managed');
    const { objects } = managedConfig;
    return findManagedObject(objects, managedObjectName) || {};
  } catch {
    return {};
  }
}

/**
 * Gets managed object properties
 *
 * @param {String} managedObjectName the managed object name used to query the resource
 * @returns {Array} list of managed object property names
 */
async function getManagedResourceProperties(managedObjectName, queryFilter = true, pageSize = -1, fields = '') {
  const { name: resourceName } = await getManagedObject(managedObjectName);

  if (!resourceName) {
    throw new Error(`Cannot find managed object: ${managedObjectName}`);
  }

  try {
    const { data } = await getManagedResourceList(resourceName, { queryFilter, pageSize, fields });
    return data.result || [];
  } catch {
    return [];
  }
}

/**
 * Gets managed object properties and sets them to corresponding data model
 * @param {Object} obj An object for the request query strings.
 * @param {String} obj.entity The entity to query.
 * @param {String} obj.attribute The attribute to query.
 * @param {String} term The search term to query.
 * @returns {Array} list of managed object property names
 */
export function managedResourcePropertyRequest({ entity, attribute }, term) {
  try {
    const queryFilter = term ? `${attribute} sw "${term}"` : true;
    return getManagedResourceProperties(entity, queryFilter, 10, attribute);
  } catch {
    return [];
  }
}

/**
 * Gets the list of all trees
 * @param {Object} config the _REPORT_FIELDS_CONTROLLER config for the corresponding reportConfig paremeter
 * @param {String} queryFilter api parameter for filtering results
 * @param {String} pageSize api parameter for returning a determined amount of results
 * @returns {Array} list of tree names
 */
export async function requestTrees(config, queryFilter, pageSize) {
  try {
    const { data } = await actionGetAllTrees(config.attribute, queryFilter, pageSize);
    return data.result || [];
  } catch {
    return [];
  }
}

/**
 * Gets a list of all certification template names
 * @param {Object} obj An object for the request query strings.
 * @param {String} obj.attribute The attribute to query.
 * @param {String} term The search term to query.
 * @returns {Array} List of governance certification templates
 */
export async function getCertificationsForReports({ attribute }, term) {
  try {
    const defaultParams = {
      queryString: term,
      pageSize: 10,
      pageNumber: 0,
      sortBy: attribute,
    };
    const { data } = await getCertificationTemplates(term, defaultParams);
    return data.result || [];
  } catch {
    return [];
  }
}

/**
 * Gets the Journey's config
 * @returns {Object} journey's config
 */
export async function getAMTreesConfig() {
  try {
    const { data } = await getAMConfig('authentication');
    return data.trees || [];
  } catch {
    return [];
  }
}

/**
 * Queries the governance resource
 * @param {Object} obj An object for the request payload.
 * @param {Function} obj.body Body function that takes in a string term.
 * @param {Object} obj.queryParameters Query parameters for the request.
 * @param {String} term The search term to query.
 * @returns {Array} List of governance resources
 */
export async function getGovernanceResource({ body, queryParameters }, term) {
  try {
    const { data } = await searchGovernanceResource(body(term), queryParameters);
    return data.result || [];
  } catch {
    return [];
  }
}

/**
 * Gets the list of governance entitlements
 * @param {Undefined} _ Unused first parameter
 * @param {String} term The search term to query.
 * @returns {Array} List of governance entitlements
 */
export async function getRequestTypesForReports(_, term) {
  try {
    const { data } = await getRequestTypes({}, term);
    return data.result || [];
  } catch {
    return [];
  }
}

/**
 * Gets the list of governance policies
 * @param {Object} obj An object for the request query strings.
 * @param {String} obj.attribute The attribute to query.
 * @param {String} term The search term to query.
 * @returns {Array} List of governance policies
 */
export async function getGovPolicyList({ attribute }, term) {
  const queryFilter = term ? `${attribute} sw "${term}"` : true;
  try {
    const { data } = await getPolicyList({ queryFilter, pageSize: 10 });
    return data.result || [];
  } catch {
    return [];
  }
}

/**
 * Gets the list of governance policy rules
 * @param {Object} obj An object for the request query strings.
 * @param {String} obj.attribute The attribute to query.
 * @param {String} term The search term to query.
 * @returns {Array} List of governance policy rules
 */
export async function getGovPolicyRuleList({ attribute }, term) {
  const queryFilter = term ? `${attribute} sw "${term}"` : true;
  try {
    const { data } = await getPolicyRuleList({ queryFilter, pageSize: 10 });
    return data.result || [];
  } catch {
    return [];
  }
}
