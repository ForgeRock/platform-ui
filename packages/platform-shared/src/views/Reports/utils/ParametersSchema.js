/**
 * Copyright (c) 2025 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { FIELD_TYPE_MAP } from './constants';
import DATA_CONTROLLER from './DataController';
import ReportParameter from './ReportParameter';
import store from '@/store';

/**
  * @description Defines the schema for the report parameters.
  * @param {Object} reportConfig Report template config parameters
  * @param {Boolean} isPrePackagedReport Determines if the report is pre-packaged (out-of-the-box)
  */

export default function ParametersSchema(reportConfig, isPrePackagedReport) {
  const isPlatformAdmin = store.state.SharedStore.currentPackage === 'admin';
  const { parameters: configParameters } = reportConfig;
  const configParameterKeys = configParameters ? Object.keys(configParameters) : [];

  if (configParameterKeys.length) {
    return configParameterKeys.map((_key) => {
      const {
        attribute,
        description = '',
        entity = '',
        enum: enums,
        label = _key,
        source,
        type = 'string',
      } = configParameters[_key];
      const SERVICE = Object.keys(DATA_CONTROLLER).find((key) => (
        Object.keys(DATA_CONTROLLER[key]).includes(entity)
      ));
      const entityOverrides = SERVICE ? DATA_CONTROLLER[SERVICE][entity] : {};
      const attributeOverrides = entityOverrides[attribute] || {};
      const { canQuery, endUserCanFetch } = entityOverrides;
      const hasEnums = !!enums?.length;
      const hasTaggableOverride = attributeOverrides.taggable === true || entityOverrides.taggable === true;
      const isSelectField = type === 'string' && (hasEnums || source === 'datasource');
      const canFetch = !!(SERVICE && !hasEnums && (isPlatformAdmin || endUserCanFetch));
      const fieldType = isSelectField ? 'select' : FIELD_TYPE_MAP[type];
      const fieldIsADropdown = fieldType === 'select' || fieldType === 'multiselect';
      const internalSearch = fieldIsADropdown && (canFetch === false || canQuery === false);
      const isUnmappedEntity = !SERVICE && source === 'datasource' && !hasEnums;
      const sourceIsBasicOrUserProvided = source === 'basic' || source === 'user_provided';
      const taggable = fieldIsADropdown && (sourceIsBasicOrUserProvided || !isPlatformAdmin || hasTaggableOverride || isUnmappedEntity);

      let model = '';
      let component = 'FrField';

      if (fieldType === 'boolean') model = false;
      if (fieldType === 'multiselect') model = [];
      if (fieldType === 'number') model = null;
      // The only real hard-coded exception here is for original out-of-the-box
      // reports needing to map to a custom component for the date range.
      if (isPrePackagedReport && _key === 'startDate') component = 'FrTimeframeField';
      if (isPrePackagedReport && _key === 'endDate') component = null;

      return {
        [_key]: new ReportParameter({
          _id: _key,
          component,
          description,
          internalSearch,
          label,
          model,
          taggable,
          type: fieldType,
          // conditional properties
          ...(hasEnums && { enums }),
          ...(canFetch && {
            request: {
              attribute,
              entity,
              data: [],
              fetch: DATA_CONTROLLER[SERVICE].fetch,
              mutation: (data, attr) => data.map((obj) => obj[attr] || obj),
              ...entityOverrides,
              ...attributeOverrides,
            },
          }),
        }),
      };
    }).reduce((a, c) => ({ ...a, ...c }), {}); // Merges the array of objects into a single object
  }
  return {};
}
