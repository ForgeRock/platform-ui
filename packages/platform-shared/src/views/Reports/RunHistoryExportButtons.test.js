/**
 * Copyright (c) 2023 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import { mount } from '@vue/test-utils';
import { findByTestId } from '@forgerock/platform-shared/src/utils/testHelpers';
import i18n from '@/i18n';
import RunHistoryExportButtons from './RunHistoryExportButtons';

describe('Run History Export Buttons component', () => {
  function setup(props) {
    return mount(RunHistoryExportButtons, {
      global: {
        plugins: [i18n],
      },
      props,
    });
  }

  let wrapper;

  describe('@renders', () => {
    it('ensures that on mount, the export icon is default', () => {
      wrapper = setup();
      const exportIcon = findByTestId(wrapper, 'fr-export-button-icon');
      expect(exportIcon.text()).toBe('sync');
    });

    it('ensures that when the exportStatus prop equals to "exporting", that the spinner icon shows', () => {
      wrapper = setup({ exportStatus: 'exporting' });

      const exportIcon = findByTestId(wrapper, 'fr-export-button-icon');
      expect(exportIcon.exists()).toBe(false);

      const spinnerIcon = findByTestId(wrapper, 'fr-export-button-spinner');
      expect(spinnerIcon.exists()).toBe(true);
    });

    it('ensures that when the exportStatus prop equals to "download", that the download icon shows', () => {
      wrapper = setup({ exportStatus: 'download' });

      const downloadIcon = findByTestId(wrapper, 'fr-download-button-icon');
      expect(downloadIcon.text()).toBe('file_download');
    });

    it('ensures that when the exportStatus prop equals to "downloading", that the download icon shows', () => {
      wrapper = setup({ exportStatus: 'downloading' });

      const spinnerButton = findByTestId(wrapper, 'fr-export-button-spinner');
      expect(spinnerButton.exists()).toBe(true);
    });

    it('ensures that when the exportStatus prop equals to "error", that the error icon shows', () => {
      wrapper = setup({ exportStatus: 'error' });

      const errorBadge = findByTestId(wrapper, 'fr-error-report-badge');
      expect(errorBadge.text()).toBe('Error');
    });

    it('ensures that the label displays when the "label" prop has a value', () => {
      wrapper = setup({ label: 'JSON' });

      const exportButtonLabelElement = findByTestId(wrapper, 'action-button-label');
      expect(exportButtonLabelElement.text()).toBe('JSON');
    });
  });
});
