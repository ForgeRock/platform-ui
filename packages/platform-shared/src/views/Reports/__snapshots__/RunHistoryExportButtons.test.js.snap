// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run History Export Buttons component @renders ensures that on mount, the export icon is default: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation RENDER_FUNCTION) Vue 3's render function API has changed. You can opt-in to the new API with:

  configureCompat({ RENDER_FUNCTION: false })

  (This can also be done per-component via the "compatConfig" option.)
  Details: https://v3-migration.vuejs.org/breaking-changes/render-function-api.html",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`Run History Export Buttons component @renders ensures that when the exportStatus prop equals to "error", that the error icon shows: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (3)",
    "
",
    " at <RunReportBadges",
    "key=0",
    "report-status="error"",
    ">",
    "
",
    " at <RunHistoryExportButtons",
    "exportStatus="error"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) (2)",
    "
",
    " at <RunReportBadges",
    "key=0",
    "report-status="error"",
    ">",
    "
",
    " at <RunHistoryExportButtons",
    "exportStatus="error"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) (2)",
    "
",
    " at <BBadge",
    "key=3",
    "class="w-100"",
    "data-testid="fr-error-report-badge"",
    " ...",
    ">",
    "
",
    " at <RunReportBadges",
    "key=0",
    "report-status="error"",
    ">",
    "
",
    " at <RunHistoryExportButtons",
    "exportStatus="error"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;

exports[`Run History Export Buttons component @renders ensures that when the exportStatus prop equals to "exporting", that the spinner icon shows: console-warn-snapshot 1`] = `
[
  [
    "[Vue warn]: (deprecation GLOBAL_EXTEND) (2)",
    "
",
    " at <RunHistoryExportButtons",
    "exportStatus="exporting"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation COMPONENT_FUNCTIONAL) Functional component <BSpinner> should be defined as a plain function in Vue 3. The "functional" option has been removed. NOTE: Before migrating to use plain functions for functional components, first make sure that all async components usage have been migrated and its compat behavior has been disabled.
  Details: https://v3-migration.vuejs.org/breaking-changes/functional-components.html",
    "
",
    " at <RunHistoryExportButtons",
    "exportStatus="exporting"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
  [
    "[Vue warn]: (deprecation INSTANCE_LISTENERS) vm.$listeners has been removed. In Vue 3, parent v-on listeners are included in vm.$attrs and it is no longer necessary to separately use v-on="$listeners" if you are already using v-bind="$attrs". (Note: the Vue 3 behavior only applies if this compat config is disabled)
  Details: https://v3-migration.vuejs.org/breaking-changes/listeners-removed.html",
    "
",
    " at <BSpinner",
    "key=1",
    "class="mr-1"",
    "data-testid="fr-export-button-spinner"",
    " ...",
    ">",
    "
",
    " at <RunHistoryExportButtons",
    "exportStatus="exporting"",
    "ref="VTU_COMPONENT"",
    ">",
    "
",
    " at <VTUROOT>",
  ],
]
`;
